plugins/markdown/core/src/org/intellij/plugins/markdown/ui/preview/jcef/CodeFenceLanguageParsingSupport.kt

```diff
@@ -0,0 +1,279 @@
+        }
+        catch (_: Exception) {}
+        scriptLatch.countDown()
+      }.start()
+
+      Thread {
+        try {
+          fallbackJS = ::CodeFenceLanguageParsingSupport.javaClass
+            .getResourceAsStream("highlighter.js")?.bufferedReader().use { it?.readText() }
+        }
+        catch (_: Exception) {}
+        scriptLatch.countDown()
+      }.start()
+
+      scriptLatch.await(5, TimeUnit.SECONDS)
+
+      val highlightJS = latestJS ?: fallbackJS
+
+      if (highlightJS == null) {
+        available = false
+        return
+      }
+
+      @Suppress("JSUnresolvedReference")
+      @Language("JavaScript")
+      val jsExtra = """
+        |function markdownHighlighter(language, content) {
+        |  let html = '';
+        |
+        |  if (language === 'json5')
+        |    language = 'javascript';
+        |
+        |  if (language) {
+        |    try {
+        |      html = hljs.highlight(content, { language }).value;
+        |    }
+        |    catch {}
+        |  }
+        |
+        |  if (!html) {
+        |    try {
+        |      html = hljs.highlightAuto(content).value;
+        |    }
+        |    catch {}
+        |  }
+        |
+        |  console.info('highlighter:' + html);
+        |}
+        """.trimMargin()
+
+      browser.createImmediately()
+      browser.waitForPageLoad("about:blank")
+      browser.executeJavaScript(highlightJS)
+      browser.executeJavaScript(jsExtra)
+      startupCompleted = true
+      startupLatch?.countDown()
+    }
+
+    private val entities = mapOf("amp" to "&", "gt" to ">", "lt" to "<", "quot" to "\"")
+
+    private fun decodeEntities(text: String): String {
+      return text.replace(Regex("&(amp|gt|lt|quot);")) { matchResult -> entities[matchResult.groupValues[1]] ?: "" }
+    }
+
+    private val reverseEntities = mapOf("&" to "&amp;", ">" to "&gt;", "<" to "&lt;", "\"" to "&quot;")
+
+    private fun encodeEntities(text: String): String {
+      return text.replace(Regex("""[&><"]""")) { matchResult -> reverseEntities[matchResult.value] ?: "" }
+    }
+
+    private val escapees: Map<Char, String> =
+      mapOf('\\' to "\\\\",  '\'' to "\\'", '"' to "\\\"", '\n' to "\\n", '\r' to "\\r", '\t' to "\\t")
+
+    private fun escapeForJs(text: String): String {
+      return text.replace(Regex("""([\\'"\n\r])""")) { match -> escapees[match.groupValues[1][0]].orEmpty() }
+        .replace(Regex("""([\x00-\x1F])""")) { match -> "\\x" + match.groupValues[1][0].code.toString(16).padStart(2, '0') }
+    }
+
+    private data class TagInfo(val line: Int, val offset: Int)
+
+    private fun addSourceRange(tag: String, start: Int, end: Int): String {
+      return tag.replace(Regex("^(<\\w+)")) { match -> "${match.value} md-src-pos=\"$start..$end\"" }
+    }
+
+    private fun convertToRangedSpans(html: String, startOffset: Int): String {
+      val chunks = html.split(Regex("""(?<=(>|\r\n|\r|\n))|(?=(<|\r\n|\r|\n))""")).toTypedArray()
```

Comment:
Compiling regex is an expensive operation

Reply:
Doesn't the compiler optimize this so the regexes only get compiled once? Or is it necessary to define these regexes separately somewhere else so to insure the regexes don't get recompiled each time they're used?

Reply:
As an experiment, I rewrote `escapeForJs()` like this, benchmarking 1000 repetitions of calling the function with assorted strings:

```kotlin
    private val basicEscapes = Regex("""([\\'"\n\r])""")
    private val controlEscapes = Regex("""([\x00-\x1F])""")

    private fun escapeForJs(text: String): String {
      return text.replace(basicEscapes) { match -> escapees[match.groupValues[1][0]].orEmpty() }
        .replace(controlEscapes) { match -> "\\x" + match.groupValues[1][0].code.toString(16).padStart(2, '0') }
    }
 ```
 
This did not perform any faster than the original version, indicating the Regexes are likely only compiled once, not repeatedly for each call.

Reply:
AFAIK, C# performs optimization, but the JVM does not.

Could I please see the documentation that explains this difference? I can't find it; stackoverflow only has outdated material.

Reply:
Even though I didn't see a measurable speed difference, you're definitely right that these Regexes would get recompiled each time, so I've fixed that problem. I've gotten spoiled by JavaScript regex literals.

Reply:
Simple iterations are good for the compiler to optimize. In this particular case, 1000 times is too few parameters. At least 100,000 iterations are recommended.